modules/text_server_adv/text_server_adv.cpp

```diff
@@ -1658,6 +1658,123 @@ float TextServerAdvanced::shaped_text_tab_align(RID p_shaped, const Vector<float
+	_THREAD_SAFE_METHOD_
+	ShapedTextDataAdvanced *sd = shaped_owner.getornull(p_shaped_line);
+	ERR_FAIL_COND_MSG(!sd, "ShapedTextDataAdvanced invalid.");
+	if (!sd->valid) {
+		shaped_text_shape(p_shaped_line);
+	}
+
+	bool add_ellipsis = (p_clip_flags & OVERRUN_ADD_ELLIPSIS) == OVERRUN_ADD_ELLIPSIS;
+	bool cut_per_word = (p_clip_flags & OVERRUN_TRIM_WORD_ONLY) == OVERRUN_TRIM_WORD_ONLY;
+	bool enforce_ellipsis = (p_clip_flags & OVERRUN_ENFORCE_ELLIPSIS) == OVERRUN_ENFORCE_ELLIPSIS;
+
+	Glyph *sd_glyphs = sd->glyphs.ptrw();
+
+	if (sd_glyphs == nullptr || p_width <= 0 || !(sd->width > p_width || enforce_ellipsis)) {
+		return;
+	}
+
+	int sd_size = sd->glyphs.size();
+
+	RID last_gl_font_rid = sd_glyphs[sd_size - 1].font_rid;
+	int last_gl_font_size = sd_glyphs[sd_size - 1].font_size;
+	uint32_t dot_gl_idx = font_get_glyph_index(last_gl_font_rid, '.');
+	Vector2 dot_adv = font_get_glyph_advance(last_gl_font_rid, dot_gl_idx, last_gl_font_size);
+	uint32_t whitespace_gl_idx = font_get_glyph_index(last_gl_font_rid, ' ');
+	Vector2 whitespace_adv = font_get_glyph_advance(last_gl_font_rid, whitespace_gl_idx, last_gl_font_size);
+
+	int ellipsis_advance;
+	if (add_ellipsis) {
+		ellipsis_advance = 3 * dot_adv.x + font_get_spacing_glyph(last_gl_font_rid) + (cut_per_word ? whitespace_adv.x : 0);
+	} else {
+		ellipsis_advance = 0;
+	}
+
+	int ell_min_characters = 6;
+	float width = sd->width;
+
+	int clip_pos = -1;
+	int ellipsis_pos = -1;
+	if (enforce_ellipsis) {
+		clip_pos = 0;
+		ellipsis_pos = 0;
+	}
+
+	int last_valid_cut = 0;
+	bool found = false;
+
+	for (int i = sd_size - 1; i >= 0; --i) {
+		if (sd_glyphs[i].count > 0) {
+			if (width - sd_glyphs[i].advance + ((i >= ell_min_characters && add_ellipsis || enforce_ellipsis) ? ellipsis_advance : 0) <= p_width) {
+				if (cut_per_word && i >= ell_min_characters) {
+					if ((sd_glyphs[i].flags & GRAPHEME_IS_BREAK_SOFT) == GRAPHEME_IS_BREAK_SOFT) {
+						last_valid_cut = i;
+						found = true;
+					}
+				} else {
+					last_valid_cut = i;
+					found = true;
+				}
+				if (found) {
+					clip_pos = last_valid_cut;
+
+					if (i >= ell_min_characters && width - ellipsis_advance <= p_width) {
+						ellipsis_pos = clip_pos;
+					}
+					break;
+				}
+			}
+		}
+
+		width -= sd_glyphs[i].advance;
+	}
+
+	if ((clip_pos >= 0 && sd->width > p_width) || enforce_ellipsis) {
+		// Cut the unwanted glyphs off.
+		sd->glyphs.resize(clip_pos);
+		sd_size = sd->glyphs.size();
+
+		if (add_ellipsis && (ellipsis_pos > 0 || enforce_ellipsis)) {
+			// Insert an additional space when cutting word bound for aesthetics.
+			if (cut_per_word && (ellipsis_pos > 0)) {
+				TextServer::Glyph gl;
+				gl.start = sd_glyphs[ellipsis_pos].start;
+				gl.end = sd_glyphs[ellipsis_pos].end;
+				gl.count = 1;
+				gl.advance = whitespace_adv.x;
+				gl.index = whitespace_gl_idx;
+				gl.font_rid = last_gl_font_rid;
+				gl.font_size = last_gl_font_size;
+				gl.flags = GRAPHEME_IS_SPACE | GRAPHEME_IS_BREAK_SOFT;
```

Comment:
The added glyphs probably should have `GRAPHEME_IS_VIRTUAL` bit set (to be non-selectable and skipped by caret movement), irrelevant for the label, but good for consistency.

Reply:
Done.